🏗️ feat!: move to using pipenv for developers and other refinements

ChasNelson1990 opened the pull request:
New features:

- Uses Python 3.9.7
- Pins dependency versions and locks
- Adds three pipenv scripts for clearing all outputs, checking formatting of and actually formatting python code in notebooks with black
- A small update to the readme for developers (us!)
- New GitHub Actions that will check you have cleared outputs and will checking the code formatting when PRs are createed
- Flake8 config file if you can get jupyterlab_flake8 working (I wasn't having much luck in Jupyter Lab but VSCode will obey this config too)
- Add pyproject.toml, at the moment just for black config
- Removed old slides notebook
- The rest should be formatting and lint corrections

closes #34 #39 #40

relates #37 - removes the old slides folder in favour of us moving the notebooks towards a more slide-y state

relates #35 - this PR sticks with Python 3.9.7 but we can easily update the Pipfile and relock when Anaconda bumps their Python version in the future

ChasNelson1990 committed:
🙈 add vscode to gitignore

ChasNelson1990 committed:
📌 move to pipenv

ChasNelson1990 committed:
📝 update readme with a line or two on pipenv

ChasNelson1990 committed:
chore: clear outputs and format notebooks

ChasNelson1990 committed:
📝 update readme with a line on running jupyter from pipenv

mixmixmix commented:
I'd rather leave a branch for 3.9 with all relevant videos (and piplock?) before upgrading. People will keep running into "old" python for a long times as most of the other tutorials and university clusters etc. will not be updated. Otherwise we could keep links/comments referencing old and new error messages?

ChasNelson1990 committed:
👷 add outputs and formatting action

ChasNelson1990 committed:
🎨 clear outputs for slides (caught by action!)

ChasNelson1990 committed:
💚 correct action for notebooks

ChasNelson1990 committed:
🎨 correct formatting issues caught by new action

ChasNelson1990 committed:
💚 move action to using pip to install pipenv

ChasNelson1990 committed:
🎨 fix whitespace lint errors caught by action

ChasNelson1990 committed:
🙈 feat: ignore rendered html files

ChasNelson1990 committed:
⬆️ chore: update dependencies

ChasNelson1990 committed:
🔥 feat!: remove old slides

ChasNelson1990 committed:
🐛  fix: remove pipenv scripts and update README

ChasNelson1990 committed:
🚨 feat: slightly change 9.1 to remove lint error

ChasNelson1990 committed:
👷 feat: point flake8 action to ChasNelson1990 commit to deal with display() warnings

ChasNelson1990 committed:
👷 feat: move to custom flake8 action

ChasNelson1990 committed:
👷 feat: make format and lint run in parallel

ChasNelson1990 committed:
🔥 feat: remove flake8 action
it was just causing to many changes in the exercises because a lot of exercises are purposefully broken

mixmixmix commented:
I should have time to test it this coming week.

mixmixmix commented:
Jupyter lab doesn't work for me 
```
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/mix/.local/share/virtualenvs/python-zero-to-hero-beginners-course-qnL34V64/lib/python3.9/site-packages/ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "/Users/mix/.local/share/virtualenvs/python-zero-to-hero-beginners-course-qnL34V64/lib/python3.9/site-packages/traitlets/config/application.py", line 975, in launch_instance
    app.initialize(argv)
  File "/Users/mix/.local/share/virtualenvs/python-zero-to-hero-beginners-course-qnL34V64/lib/python3.9/site-packages/traitlets/config/application.py", line 110, in inner
    return method(app, *args, **kwargs)
  File "/Users/mix/.local/share/virtualenvs/python-zero-to-hero-beginners-course-qnL34V64/lib/python3.9/site-packages/ipykernel/kernelapp.py", line 681, in initialize
    self.init_kernel()
  File "/Users/mix/.local/share/virtualenvs/python-zero-to-hero-beginners-course-qnL34V64/lib/python3.9/site-packages/ipykernel/kernelapp.py", line 528, in init_kernel
    kernel = kernel_factory(
  File "/Users/mix/.local/share/virtualenvs/python-zero-to-hero-beginners-course-qnL34V64/lib/python3.9/site-packages/traitlets/config/configurable.py", line 551, in instance
    inst = cls(*args, **kwargs)
  File "/Users/mix/.local/share/virtualenvs/python-zero-to-hero-beginners-course-qnL34V64/lib/python3.9/site-packages/ipykernel/ipkernel.py", line 113, in __init__
    import appnope
ModuleNotFoundError: No module named 'appnope'
```
I will investigate

mixmixmix committed:
Added appnope required by jupyter lab on macos

mixmixmix reviewed (commented):
👍
